File: checkmath_app/lib/core/game_logic.dart
/// SciDAMA / CheckMath game engine — pure Dart, no network required.
///
/// Rules:
/// - Play squares: (row + col) % 2 == 1
/// - Row 0 = bottom; human pieces start on rows 0-2, AI on rows 5-7
/// - Captures are OPTIONAL on the first move of a turn (strategic choice)
/// - Once a capture chain is started, the player MUST continue if more
///   captures are available from the landing square (chain-continuation)
/// - SciDama scoring: the landing square's operator is applied between
///   the taker's value and the captured piece's value
library;

import 'dart:math';

import 'board_data.dart';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------
const int kEmpty = 0;
const int kHumanMan = 1;
const int kHumanKing = 2;
const int kAiMan = 3;
const int kAiKing = 4;

// ---------------------------------------------------------------------------
// Piece helpers
// ---------------------------------------------------------------------------
int pieceType(int p) => p > 0 ? p ~/ 100 : 0;
int pieceValue(int p) => p > 0 ? p % 100 : 0;
int makePiece(int type, int value) => type * 100 + value;

bool isPlaySquare(int r, int c) =>
    r >= 0 && r < 8 && c >= 0 && c < 8 && (r + c) % 2 == 1;

bool isHumanPiece(int p) {
  final t = pieceType(p);
  return t == kHumanMan || t == kHumanKing;
}

bool isAiPiece(int p) {
  final t = pieceType(p);
  return t == kAiMan || t == kAiKing;
}

bool isKing(int p) {
  final t = pieceType(p);
  return t == kHumanKing || t == kAiKing;
}

/// Returns diagonal direction vectors for a piece.
List<List<int>> directionsFor(int p, bool human) {
  if (isKing(p)) return [[-1, -1], [-1, 1], [1, -1], [1, 1]];
  return human ? [[1, -1], [1, 1]] : [[-1, -1], [-1, 1]];
}

int maybePromote(int r, int c, int p) {
  final t = pieceType(p);
  final v = pieceValue(p);
  if (t == kHumanMan && r == 7) return makePiece(kHumanKing, v);
  if (t == kAiMan && r == 0) return makePiece(kAiKing, v);
  return p;
}

// ---------------------------------------------------------------------------
// Board initialisation
// ---------------------------------------------------------------------------
List<List<int>> initialBoard() {
  final rng = Random();
  final b = List.generate(8, (_) => List.filled(8, kEmpty));
  final humanVals = List<int>.generate(12, (i) => i + 1)..shuffle(rng);
  final aiVals = List<int>.generate(12, (i) => i + 1)..shuffle(rng);
  int hIdx = 0, aIdx = 0;
  for (int r = 0; r < 8; r++) {
    for (int c = 0; c < 8; c++) {
      if (!isPlaySquare(r, c)) continue;
      if (r < 3) {
        b[r][c] = makePiece(kHumanMan, humanVals[hIdx++]);
      } else if (r > 4) {
        b[r][c] = makePiece(kAiMan, aiVals[aIdx++]);
      }
    }
  }
  return b;
}

// ---------------------------------------------------------------------------
// GameState
// ---------------------------------------------------------------------------
class GameState {
  List<List<int>> board;
  String currentTurn; // 'human' | 'ai'
  double humanScore;
  double aiScore;
  String difficulty;
  bool gameOver;
  String gameMode;
  String player1Name;
  String player2Name;
  String? winner;
  bool lastCapture; // true if last move was a capture (chain may continue)
  List<int>? lastMovedPiece; // [r, c] of piece that must continue chain
  List<String> calculations;
  bool playerCaptured;
  String gameId;

  GameState({
    required this.board,
    required this.currentTurn,
    required this.humanScore,
    required this.aiScore,
    required this.difficulty,
    required this.gameId,
    this.gameOver = false,
    this.gameMode = 'vs_ai',
    this.player1Name = 'Player 1',
    this.player2Name = 'Bot',
    this.winner,
    this.lastCapture = false,
    this.lastMovedPiece,
    List<String>? calculations,
    this.playerCaptured = false,
  }) : calculations = calculations ?? [];

  /// Convert to the same JSON shape the HTTP backend returns,
  /// so the rest of the app can consume it identically.
  Map<String, dynamic> toResponseMap() => {
        'game_id': gameId,
        'board': board,
        'current_turn': currentTurn,
        'human_score': humanScore,
        'ai_score': aiScore,
        'game_mode': gameMode,
        'player1_name': player1Name,
        'player2_name': player2Name,
        'difficulty': difficulty,
        'game_over': gameOver,
        'winner': winner,
        'last_capture': lastCapture,
        'player_captured': playerCaptured,
        'board_symbols': kBoardSymbols,
        'calculations': calculations,
      };

  /// Serialise to a plain Map for use with Flutter's compute() isolate.
  Map<String, dynamic> toIsolateMap() => {
        'board': board.map((r) => List<int>.from(r)).toList(),
        'current_turn': currentTurn,
        'human_score': humanScore,
        'ai_score': aiScore,
        'difficulty': difficulty,
        'game_over': gameOver,
        'game_mode': gameMode,
        'player1_name': player1Name,
        'player2_name': player2Name,
        'winner': winner,
        'last_capture': lastCapture,
        'last_moved_piece': lastMovedPiece,
        'calculations': List<String>.from(calculations),
        'player_captured': playerCaptured,
        'game_id': gameId,
      };

  /// Reconstruct from an isolate map.
  factory GameState.fromIsolateMap(Map<String, dynamic> m) {
    final rawBoard = m['board'] as List;
    final board = rawBoard
        .map((row) => List<int>.from((row as List).map((e) => e as int)))
        .toList();
    final rawLmp = m['last_moved_piece'];
    return GameState(
      board: board,
      currentTurn: m['current_turn'] as String,
      humanScore: (m['human_score'] as num).toDouble(),
      aiScore: (m['ai_score'] as num).toDouble(),
      difficulty: m['difficulty'] as String,
      gameOver: m['game_over'] as bool,
      gameMode: m['game_mode'] as String,
      player1Name: m['player1_name'] as String,
      player2Name: m['player2_name'] as String,
      winner: m['winner'] as String?,
      lastCapture: m['last_capture'] as bool,
      lastMovedPiece: rawLmp != null
          ? List<int>.from((rawLmp as List).map((e) => e as int))
          : null,
      calculations: List<String>.from(m['calculations'] as List),
      playerCaptured: m['player_captured'] as bool,
      gameId: m['game_id'] as String,
    );
  }

  GameState copy() => GameState(
        board: board.map((row) => List<int>.from(row)).toList(),
        currentTurn: currentTurn,
        humanScore: humanScore,
        aiScore: aiScore,
        difficulty: difficulty,
        gameOver: gameOver,
        gameMode: gameMode,
        player1Name: player1Name,
        player2Name: player2Name,
        winner: winner,
        lastCapture: lastCapture,
        lastMovedPiece:
            lastMovedPiece != null ? List<int>.from(lastMovedPiece!) : null,
        calculations: List<String>.from(calculations),
        playerCaptured: playerCaptured,
        gameId: gameId,
      );
}

// ---------------------------------------------------------------------------
// SciDAMA scoring
// ---------------------------------------------------------------------------
double calculateScore(double taker, String symbol, double taken) {
  switch (symbol) {
    case '+':
      return taker + taken;
    case '-':
      return taker - taken;
    case '×':
    case 'x':
    case 'X':
    case '*':
      return taker * taken;
    case '÷':
    case '/':
      return taken != 0 ? taker / taken : taker;
    default:
      return taker;
  }
}

// ---------------------------------------------------------------------------
// Move generation
// ---------------------------------------------------------------------------

/// Returns all multi-jump capture paths reachable from (r, c) in one sequence.
List<List<Map<String, dynamic>>> getAllPathsFrom(
  List<List<int>> board,
  int r,
  int c,
  bool human,
) {
  final p = board[r][c];
  final dirs = directionsFor(p, human);
  final paths = <List<Map<String, dynamic>>>[];

  for (final d in dirs) {
    final dr = d[0] * 2;
    final dc = d[1] * 2;
    final nr = r + dr;
    final nc = c + dc;
    if (nr < 0 || nr >= 8 || nc < 0 || nc >= 8) continue;
    if (board[nr][nc] != kEmpty) continue;

    final midR = (r + nr) ~/ 2;
    final midC = (c + nc) ~/ 2;
    final cap = board[midR][midC];
    final isEnemy = human ? isAiPiece(cap) : isHumanPiece(cap);
    if (!isEnemy) continue;

    final move = <String, dynamic>{
      'from': [r, c],
      'to': [nr, nc],
      'capture': true,
      'captured_r': midR,
      'captured_c': midC,
    };

    // Simulate capture and recurse
    final nb = board.map((row) => List<int>.from(row)).toList();
    nb[nr][nc] = nb[r][c];
    nb[r][c] = kEmpty;
    nb[midR][midC] = kEmpty;

    final subPaths = getAllPathsFrom(nb, nr, nc, human);
    if (subPaths.isEmpty) {
      paths.add([move]);
    } else {
      for (final sub in subPaths) {
        paths.add([move, ...sub]);
      }
    }
  }
  return paths;
}

/// Returns all legal moves for the given side.
///
/// **Optional-capture rule (player-facing):**
/// When [mustStartAt] is null (first move of a turn), both sliding moves
/// AND capture moves are returned — the player is NOT forced to capture.
///
/// **Chain-capture enforcement:**
/// When [mustStartAt] is set (mid-sequence), only capture continuations
/// from that specific square are returned, enforcing the chain.
///
/// The AI internally uses this same function, so it sees all options and
/// can greedily/optimally choose captures via its evaluation logic.
List<Map<String, dynamic>> allLegalMoves(
  List<List<int>> board,
  bool human, {
  List<int>? mustStartAt,
}) {
  final moves = <Map<String, dynamic>>[];

  // --- Mid-capture chain: must continue from the piece that already jumped ---
  if (mustStartAt != null) {
    final r = mustStartAt[0];
    final c = mustStartAt[1];
    final capPaths = getAllPathsFrom(board, r, c, human);
    for (final path in capPaths) {
      final mv = path[0];
      final f = mv['from'] as List;
      final t = mv['to'] as List;
      if (!moves.any((m) =>
          (m['from'] as List)[0] == f[0] &&
          (m['from'] as List)[1] == f[1] &&
          (m['to'] as List)[0] == t[0] &&
          (m['to'] as List)[1] == t[1])) {
        moves.add(mv);
      }
    }
    return moves;
  }

  // --- Normal turn: ALL moves are available (sliding + captures) ---
  for (int r = 0; r < 8; r++) {
    for (int c = 0; c < 8; c++) {
      final p = board[r][c];
      if (human && !isHumanPiece(p)) continue;
      if (!human && !isAiPiece(p)) continue;

      // Capture moves (added first so UI / AI can detect them easily)
      final capPaths = getAllPathsFrom(board, r, c, human);
      for (final path in capPaths) {
        final mv = path[0];
        final f = mv['from'] as List;
        final t = mv['to'] as List;
        if (!moves.any((m) =>
            (m['from'] as List)[0] == f[0] &&
            (m['from'] as List)[1] == f[1] &&
            (m['to'] as List)[0] == t[0] &&
            (m['to'] as List)[1] == t[1])) {
          moves.add(mv);
        }
      }

      // Sliding moves
      final dirs = directionsFor(p, human);
      for (final d in dirs) {
        final nr = r + d[0];
        final nc = c + d[1];
        if (nr >= 0 && nr < 8 && nc >= 0 && nc < 8 && board[nr][nc] == kEmpty) {
          moves.add({'from': [r, c], 'to': [nr, nc], 'capture': false});
        }
      }
    }
  }
  return moves;
}

// ---------------------------------------------------------------------------
// Move application
// ---------------------------------------------------------------------------

/// Applies a move to [state]. Returns `(ok, message)`.
/// Mutates [state] in place.
(bool, String) applyMove(
  GameState state,
  int fr,
  int fc,
  int tr,
  int tc,
  bool human,
) {
  final b = state.board;
  if (!isPlaySquare(fr, fc) || !isPlaySquare(tr, tc)) {
    return (false, 'Invalid squares');
  }
  final piece = b[fr][fc];
  if (human && !isHumanPiece(piece)) return (false, 'Not your piece');
  if (!human && !isAiPiece(piece)) return (false, 'Not your piece');

  final mustStart = state.lastCapture ? state.lastMovedPiece : null;
  final legal = allLegalMoves(b, human, mustStartAt: mustStart);

  final match = legal.firstWhere(
    (m) =>
        (m['from'] as List)[0] == fr &&
        (m['from'] as List)[1] == fc &&
        (m['to'] as List)[0] == tr &&
        (m['to'] as List)[1] == tc,
    orElse: () => {},
  );

  if (match.isEmpty) return (false, 'Illegal move');

  state.playerCaptured = match['capture'] == true;

  if (match['capture'] != true) {
    // Simple slide
    b[tr][tc] = maybePromote(tr, tc, piece);
    b[fr][fc] = kEmpty;
    state.currentTurn = human ? 'ai' : 'human';
    state.lastCapture = false;
    state.lastMovedPiece = null;
    return (true, 'ok');
  }

  // Capture
  final midR = match['captured_r'] as int;
  final midC = match['captured_c'] as int;
  final capturedPiece = b[midR][midC];

  b[midR][midC] = kEmpty;
  b[tr][tc] = maybePromote(tr, tc, piece);
  b[fr][fc] = kEmpty;

  // SciDama scoring
  final sym = kBoardSymbols[tr][tc];
  if (sym.isNotEmpty) {
    final takerVal = pieceValue(piece).toDouble();
    final takenVal = pieceValue(capturedPiece).toDouble();
    final result = calculateScore(takerVal, sym, takenVal);

    int multiplier = 1;
    String desc = '';
    if (isKing(piece) && isKing(capturedPiece)) {
      multiplier = 4;
      desc = ' (Dama×Dama x4)';
    } else if (isKing(piece) || isKing(capturedPiece)) {
      multiplier = 2;
      desc = ' (Dama ×2)';
    }

    final finalAdded = result.abs() * multiplier;

    if (human) {
      state.humanScore += finalAdded;
      state.calculations.insert(
        0,
        '👤 ${state.player1Name}: ${ takerVal.toInt()} $sym ${takenVal.toInt()} = ${result.toStringAsFixed(1)}$desc  (+${finalAdded.toStringAsFixed(1)})',
      );
    } else {
      state.aiScore += finalAdded;
      final av = state.gameMode == 'vs_ai' ? '🤖' : '👤';
      state.calculations.insert(
        0,
        '$av ${state.player2Name}: ${takerVal.toInt()} $sym ${takenVal.toInt()} = ${result.toStringAsFixed(1)}$desc  (+${finalAdded.toStringAsFixed(1)})',
      );
    }
    if (state.calculations.length > 20) state.calculations.removeLast();
  }

  // Check for chain continuation
  state.lastCapture = true;
  state.lastMovedPiece = [tr, tc];

  final nextMoves = allLegalMoves(b, human, mustStartAt: [tr, tc]);
  if (nextMoves.isNotEmpty && nextMoves.first['capture'] == true) {
    // Same side continues the chain
    state.currentTurn = human ? 'human' : 'ai';
  } else {
    state.currentTurn = human ? 'ai' : 'human';
    state.lastCapture = false;
    state.lastMovedPiece = null;
  }

  return (true, 'ok');
}

// ---------------------------------------------------------------------------
// Piece counting + game-end
// ---------------------------------------------------------------------------
int countPieces(List<List<int>> board, bool human) {
  int n = 0;
  for (int r = 0; r < 8; r++) {
    for (int c = 0; c < 8; c++) {
      final p = board[r][c];
      if (human && isHumanPiece(p)) n++;
      if (!human && isAiPiece(p)) n++;
    }
  }
  return n;
}

void checkGameEnd(GameState state) {
  final h = countPieces(state.board, true);
  final a = countPieces(state.board, false);

  bool hasLegalMoves = false;
  if (h > 0 && a > 0) {
    final legal = allLegalMoves(
      state.board,
      state.currentTurn == 'human',
      mustStartAt: state.lastCapture ? state.lastMovedPiece : null,
    );
    hasLegalMoves = legal.isNotEmpty;
  }

  if (h == 0 || a == 0 || !hasLegalMoves) {
    state.gameOver = true;

    // Add remaining chip values to their owners' scores
    for (int r = 0; r < 8; r++) {
      for (int c = 0; c < 8; c++) {
        final p = state.board[r][c];
        if (p == kEmpty) continue;
        final val = pieceValue(p).toDouble();
        final mult = isKing(p) ? 2.0 : 1.0;
        if (isHumanPiece(p)) {
          state.humanScore += val * mult;
        } else {
          state.aiScore += val * mult;
        }
      }
    }

    if (state.humanScore > state.aiScore) {
      state.winner = 'human';
    } else if (state.aiScore > state.humanScore) {
      state.winner = 'ai';
    } else {
      state.winner = 'draw';
    }
  }
}

File: backend/services/game_logic.py
"""
CheckMath: checkers on squares where (row + col) % 2 == 1 (symbol / 'white' tiles).
Row 0 is the bottom; human plays from bottom, AI from top.
SciDama rules applied: Mayor (must take max path length) and calculating based on chip values.
"""
from __future__ import annotations

import copy
import uuid
import random
from dataclasses import dataclass, field
from typing import Any, Literal, Optional, List, Dict, Tuple

from services.scoring import calculate_score

# Board layout: row 0 = bottom. Symbols on odd (row+col) cells only; even cells are "".
BOARD_SYMBOLS: list[list[str]] = [
    ["", "+", "", "-", "", "÷", "", "×"],
    ["-", "", "+", "", "×", "", "÷", ""],
    ["", "÷", "", "×", "", "+", "", "-"],
    ["×", "", "÷", "", "-", "", "+", ""],
    ["", "+", "", "-", "", "÷", "", "×"],
    ["-", "", "+", "", "×", "", "÷", ""],
    ["", "÷", "", "×", "", "+", "", "-"],
    ["×", "", "÷", "", "-", "", "+", ""],
]

EMPTY = 0
HUMAN_MAN = 1
HUMAN_KING = 2
AI_MAN = 3
AI_KING = 4

def piece_type(p: int) -> int:
    return p // 100 if p > 0 else 0

def piece_value(p: int) -> int:
    return p % 100 if p > 0 else 0

def make_piece(type_: int, value: int) -> int:
    return type_ * 100 + value

def is_play_square(r: int, c: int) -> bool:
    return 0 <= r < 8 and 0 <= c < 8 and (r + c) % 2 == 1

def initial_board() -> list[list[int]]:
    b = [[EMPTY] * 8 for _ in range(8)]
    human_vals = list(range(1, 13))
    ai_vals = list(range(1, 13))
    
    random.shuffle(human_vals)
    random.shuffle(ai_vals)
    
    h_idx = 0
    a_idx = 0
    for r in range(8):
        for c in range(8):
            if not is_play_square(r, c):
                continue
            if r < 3:
                b[r][c] = make_piece(HUMAN_MAN, human_vals[h_idx])
                h_idx += 1
            elif r > 4:
                b[r][c] = make_piece(AI_MAN, ai_vals[a_idx])
                a_idx += 1
    return b

def _is_human(p: int) -> bool:
    return piece_type(p) in (HUMAN_MAN, HUMAN_KING)

def _is_ai(p: int) -> bool:
    return piece_type(p) in (AI_MAN, AI_KING)

def _is_king(p: int) -> bool:
    return piece_type(p) in (HUMAN_KING, AI_KING)

def _directions_for_piece(p: int, side: Literal["human", "ai"]) -> list[tuple[int, int]]:
    if _is_king(p):
        return [(-1, -1), (-1, 1), (1, -1), (1, 1)]
    if side == "human":
        return [(1, -1), (1, 1)]
    return [(-1, -1), (-1, 1)]

def _maybe_promote(r: int, c: int, p: int) -> int:
    pt = piece_type(p)
    val = piece_value(p)
    if pt == HUMAN_MAN and r == 7:
        return make_piece(HUMAN_KING, val)
    if pt == AI_MAN and r == 0:
        return make_piece(AI_KING, val)
    return p

@dataclass
class GameState:
    board: list[list[int]]
    current_turn: Literal["human", "ai"]
    human_score: float
    ai_score: float
    difficulty: str
    game_over: bool = False
    game_mode: str = "vs_ai"
    player1_name: str = "Player 1"
    player2_name: str = "Bot"
    winner: Optional[Literal["human", "ai", "draw"]] = None
    last_capture: bool = False
    last_moved_piece: Optional[Tuple[int, int]] = None
    calculations: List[str] = field(default_factory=list)
    player_captured: bool = False

    def to_dict(self) -> dict[str, Any]:
        return {
            "game_id": getattr(self, "_game_id", ""),
            "board": self.board,
            "current_turn": self.current_turn,
            "human_score": self.human_score,
            "ai_score": self.ai_score,
            "game_mode": self.game_mode,
            "player1_name": self.player1_name,
            "player2_name": self.player2_name,
            "difficulty": self.difficulty,
            "game_over": self.game_over,
            "winner": self.winner,
            "last_capture": self.last_capture,
            "player_captured": self.player_captured,
            "board_symbols": BOARD_SYMBOLS,
            "calculations": self.calculations,
        }

GAMES: dict[str, GameState] = {}

def create_game(difficulty: str, game_mode: str = "vs_ai", p1: str = "Player", p2: str = "Bot", starting_turn: str = "human") -> tuple[str, GameState]:
    gid = str(uuid.uuid4())
    state = GameState(
        board=initial_board(),
        current_turn=starting_turn,
        human_score=0.0,
        ai_score=0.0,
        difficulty=difficulty,
        game_mode=game_mode,
        player1_name=p1,
        player2_name=p2,
    )
    setattr(state, "_game_id", gid)
    GAMES[gid] = state
    return gid, state

def get_game(game_id: str) -> Optional[GameState]:
    return GAMES.get(game_id)

def count_pieces(board: list[list[int]], human: bool) -> int:
    s = 0
    for r in range(8):
        for c in range(8):
            if human and _is_human(board[r][c]):
                s += 1
            if not human and _is_ai(board[r][c]):
                s += 1
    return s

# --- MAYOR RULES IMPLEMENTATION ---
def get_all_paths_from(board: list[list[int]], r: int, c: int, human: bool) -> List[List[Dict]]:
    """Returns a list of all complete capture paths starting from (r,c)."""
    p = board[r][c]
    side: Literal["human", "ai"] = "human" if human else "ai"
    dirs = _directions_for_piece(p, side)
    
    paths = []
    
    for dr in (-2, 2):
        for dc in (-2, 2):
            nr, nc = r + dr, c + dc
            if not (0 <= nr < 8 and 0 <= nc < 8):
                continue
            if board[nr][nc] != EMPTY:
                continue
            
            mid_r, mid_c = (r + nr) // 2, (c + nc) // 2
            cap = board[mid_r][mid_c]
            enemy = _is_ai(cap) if human else _is_human(cap)
            if not enemy:
                continue
            
            step_r, step_c = dr // 2, dc // 2
            if (step_r, step_c) not in dirs:
                continue
            
            # Formulate the move
            move = {
                "from": [r, c],
                "to": [nr, nc],
                "capture": True,
                "captured_r": mid_r,
                "captured_c": mid_c
            }
            
            # Recurse from new position
            new_board = [row[:] for row in board]
            new_board[nr][nc] = new_board[r][c] # Piece moves
            new_board[r][c] = EMPTY
            new_board[mid_r][mid_c] = EMPTY # enemy removed
            
            sub_paths = get_all_paths_from(new_board, nr, nc, human)
            if not sub_paths:
                paths.append([move])
            else:
                for sub in sub_paths:
                    paths.append([move] + sub)
                    
    return paths

def all_legal_moves(board: list[list[int]], human: bool, must_start_at: Optional[Tuple[int,int]] = None) -> List[Dict]:
    """
    Finds all valid FIRST steps according to Mayor Rules.
    If must_start_at is given, we ONLY consider moves starting with that piece.
    """
    moves = []
    
    # 1. Gather all capture paths
    all_capture_paths = []
    starts = [must_start_at] if must_start_at else [(r, c) for r in range(8) for c in range(8)]
    
    for coord in starts:
        if coord is None: continue
        r, c = coord
        p = board[r][c]
        if human and not _is_human(p): continue
        if not human and not _is_ai(p): continue
        
        paths = get_all_paths_from(board, r, c, human)
        all_capture_paths.extend(paths)
        
    # 2. Filter longest paths (Mayor Rules)
    if all_capture_paths:
        max_len = max(len(p) for p in all_capture_paths)
        longest_paths = [p for p in all_capture_paths if len(p) == max_len]
        
        # We only return the FIRST step of the longest paths as valid next moves
        for p in longest_paths:
            move = p[0]
            # Ensure uniqueness
            if not any(m["from"] == move["from"] and m["to"] == move["to"] for m in moves):
                moves.append(move)
        return moves

    # 3. If no captures are possible, and we are NOT forced to start at a specific piece (mid-jump)
    if must_start_at is not None:
        return [] # Mid-capture but no more captures available == shouldn't happen, but just in case
        
    # Standard sliding moves
    for r in range(8):
        for c in range(8):
            p = board[r][c]
            if human and not _is_human(p): continue
            if not human and not _is_ai(p): continue
            
            side: Literal["human", "ai"] = "human" if human else "ai"
            dirs = _directions_for_piece(p, side)
            
            for dr, dc in dirs:
                nr, nc = r + dr, c + dc
                if 0 <= nr < 8 and 0 <= nc < 8 and board[nr][nc] == EMPTY:
                    moves.append({"from": [r, c], "to": [nr, nc], "capture": False})
                    
    return moves

def try_move(
    game_id: str,
    from_r: int,
    from_c: int,
    to_r: int,
    to_c: int,
) -> tuple[bool, str, Optional[GameState]]:
    state = GAMES.get(game_id)
    if not state or state.game_over:
        return False, "Invalid or finished game", None
    if state.current_turn != "human" and state.game_mode == "vs_ai":
        return False, "Not your turn", None

    is_human = (state.current_turn == "human")
    ok, msg = _apply_move(state, from_r, from_c, to_r, to_c, is_human)
    if not ok:
        return False, msg, None

    check_game_end(state)
    return True, "ok", state
    
def apply_ai_move(state: GameState, fr: int, fc: int, tr: int, tc: int) -> bool:
    ok, _ = _apply_move(state, fr, fc, tr, tc, False)
    return ok

def _apply_move(state: GameState, fr: int, fc: int, tr: int, tc: int, human: bool) -> tuple[bool, str]:
    b = state.board
    if not (is_play_square(fr, fc) and is_play_square(tr, tc)):
        return False, "Invalid squares"
    piece = b[fr][fc]
    if human and not _is_human(piece):
        return False, "Not your piece"
    if not human and not _is_ai(piece):
        return False, "Not your piece"

    # Enforce sequence starting from last moved piece if mid-capture
    must_start_at = state.last_moved_piece if state.last_capture else None
    
    legal_starts = all_legal_moves(b, human, must_start_at)
    match = next((m for m in legal_starts if m["from"] == [fr, fc] and m["to"] == [tr, tc]), None)
    
    if not match:
        return False, "Illegal move. Mayor rule dictates maximum capture sequence!"
        
    state.player_captured = match["capture"]

    if not match["capture"]:
        # Simple move
        b[tr][tc] = _maybe_promote(tr, tc, piece)
        b[fr][fc] = EMPTY
        state.current_turn = "ai" if human else "human"
        state.last_capture = False
        state.last_moved_piece = None
        return True, "ok"

    # Capture move
    mid_r, mid_c = match["captured_r"], match["captured_c"]
    captured_piece = b[mid_r][mid_c]
    
    b[mid_r][mid_c] = EMPTY
    b[tr][tc] = _maybe_promote(tr, tc, piece)
    b[fr][fc] = EMPTY
    
    # SCI-DAMA SCORING
    sym = BOARD_SYMBOLS[tr][tc]
    taker_val = piece_value(piece)
    taken_val = piece_value(captured_piece)
    
    if sym:
        result = calculate_score(float(taker_val), sym, float(taken_val))
        
        # Multipliers
        multiplier = 1
        desc = ""
        if _is_king(piece) and _is_king(captured_piece):
            multiplier = 4
            desc = " (Dama takes Dama x4)"
        elif _is_king(piece) or _is_king(captured_piece):
            multiplier = 2
            desc = " (Dama involved x2)"
            
        final_added = result * multiplier
        
        final_added = abs(result * multiplier)
        
        if human:
            state.human_score += final_added
            log = f"👤 {state.player1_name}: {taker_val} {sym} {taken_val} = {result}{desc} (+{final_added} to Score)"
        else:
            state.ai_score += final_added
            av = "🤖" if state.game_mode == "vs_ai" else "👤"
            log = f"{av} {state.player2_name}: {taker_val} {sym} {taken_val} = {result}{desc} (+{final_added} to Score)"
            
        state.calculations.insert(0, log)
        if len(state.calculations) > 10:
            state.calculations.pop()

    # Continue sequence?
    # We must check if `all_legal_moves` with `must_start_at=(tr, tc)` has ANY captures.
    # Since we enforced Mayor from the start, we are guaranteed to be on the longest path.
    # So if there's a capture left, it's the continuation.
    state.last_capture = True
    state.last_moved_piece = (tr, tc)
    
    next_moves = all_legal_moves(b, human, must_start_at=(tr, tc))
    if next_moves and next_moves[0]["capture"]:
        state.current_turn = "human" if human else "ai" # Same player
    else:
        state.current_turn = "ai" if human else "human" # Next player
        state.last_capture = False
        state.last_moved_piece = None
        
    return True, "ok"

def check_game_end(state: GameState) -> None:
    h = count_pieces(state.board, True)
    a = count_pieces(state.board, False)
    
    legal_moves_available = False
    if h > 0 and a > 0:
        legal = all_legal_moves(state.board, state.current_turn == "human", state.last_moved_piece if state.last_capture else None)
        legal_moves_available = len(legal) > 0
        
    if h == 0 or a == 0 or not legal_moves_available:
        state.game_over = True
        
        # End game tally: remaining chips are added to their respective owners.
        # If the chip is a Dama, its value is doubled.
        for r in range(8):
            for c in range(8):
                p = state.board[r][c]
                if p == EMPTY:
                    continue
                val = piece_value(p)
                mult = 2 if _is_king(p) else 1
                if _is_human(p):
                    state.human_score += (val * mult)
                else:
                    state.ai_score += (val * mult)
                    
        # Winner is whoever has the highest score
        if state.human_score > state.ai_score:
            state.winner = "human"
        elif state.ai_score > state.human_score:
            state.winner = "ai"
        else:
            state.winner = "draw"

def copy_board(board: list[list[int]]) -> list[list[int]]:
    return copy.deepcopy(board)

def copy_state(state: GameState) -> GameState:
    return GameState(
        board=copy_board(state.board),
        current_turn=state.current_turn,
        human_score=state.human_score,
        ai_score=state.ai_score,
        difficulty=state.difficulty,
        game_mode=state.game_mode,
        player1_name=state.player1_name,
        player2_name=state.player2_name,
        game_over=state.game_over,
        winner=state.winner,
        last_capture=state.last_capture,
        last_moved_piece=state.last_moved_piece,
        calculations=state.calculations[:]
    )
